### PR TITLE
[Feature] QM9Edge Dataset Support

### PR DESCRIPTION
**Pull request by hengruizhang98**
## Description
QM9Edge Dataset Support.  #2567 
 @mufeili 

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [x] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage
- [x] Code is well-documented
- [x] To my best knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [x] Related issue is referred in this PR

### CODE REVIEW

File: python/dgl/data/qm9_edge.py [hunk 1]
```diff
@@ -0,0 +1,404 @@
+"""QM9 dataset for graph property prediction (regression)."""
+"""When building this class, we partially refer to the implementation from https://github.com/Jack-XHP/DGL_QM9EDGE """
```
Comment (mufeili) [inline python/dgl/data/qm9_edge.py:2]: You only need one pair of """, i.e.

```python
"""QM9 dataset for graph property prediction (regression).

When building this class, we partially refer to the implementation from https://github.com/Jack-XHP/DGL_QM9EDGE
"""
```

File: python/dgl/data/qm9_edge.py [hunk 2]
```diff
@@ -0,0 +1,404 @@
...
+import os
+import numpy as np
+
+from .dgl_dataset import DGLDataset
+from .utils import download, extract_archive, _get_dgl_url
+from ..convert import graph as dgl_graph
+from .. import backend as F
+
+try:
+    import rdkit
```
Comment (mufeili) [inline python/dgl/data/qm9_edge.py:13]: @jermainewang Are you fine with having RDKit as an optional dependency when importing this dataset?

Comment (mufeili) [inline python/dgl/data/qm9_edge.py:13]: rdkit not used

**Comment by mufeili** [comment]
I still find this confusing. If a user has not installed `rdkit`, then this class will use the preprocessed dataset. Then when will a user ever want to preprocess from scratch and install rdkit for doing so? If you just want to display the details of preprocessing as @BarclayII suggested, then just always let users preprocess from scratch and save the preprocessed dataset to the local disk.

File: python/dgl/data/qm9_edge.py [hunk 3]
```diff
@@ -0,0 +1,404 @@
...
+    from rdkit import Chem
+    from rdkit.Chem.rdchem import HybridizationType
+    from rdkit.Chem.rdchem import BondType as BT
+    from rdkit import RDLogger
+    RDLogger.DisableLog('rdApp.*')
+except ImportError:
+    rdkit = None
+HAR2EV = 27.2113825435
```
Comment (mufeili) [inline python/dgl/data/qm9_edge.py:21]: Add a comment to explain it

File: python/dgl/data/qm9_edge.py [hunk 4]
```diff
@@ -0,0 +1,404 @@
...
+KCALMOL2EV = 0.04336414
```
Comment (mufeili) [inline python/dgl/data/qm9_edge.py:22]: Ditto

**Comment by mufeili** [comment]
You also need to update [this file](https://github.com/dmlc/dgl/blob/master/docs/source/api/python/dgl.data.rst) to update the documentation.

File: python/dgl/data/qm9_edge.py [hunk 5]
```diff
@@ -0,0 +1,404 @@
...
+conversion = F.tensor([
+    1., 1., HAR2EV, HAR2EV, HAR2EV, 1., HAR2EV, HAR2EV, HAR2EV, HAR2EV, HAR2EV,
+    1., KCALMOL2EV, KCALMOL2EV, KCALMOL2EV, KCALMOL2EV, 1., 1., 1.
+])
+
+
+class QM9EdgeDataset(DGLDataset):
+    r"""QM9Edge dataset for graph property prediction (regression)
+    This dataset consists of 130,831 molecules with 19 regression targets.
+    Node means atom and edge means bond.
```
Comment (mufeili) [inline python/dgl/data/qm9_edge.py:32]: Nodes correspond to atoms and edges correspond to bonds.

File: python/dgl/data/qm9_edge.py [hunk 6]
```diff
@@ -0,0 +1,404 @@
...
+    
+    This dataset differs from QM9Dataset in the following points:
```
Comment (mufeili) [inline python/dgl/data/qm9_edge.py:34]: "QM9Dataset" -> ":class:`~dgl.data.QM9Dataset`" for hyperlink in rendering the documentation

File: python/dgl/data/qm9_edge.py [hunk 7]
```diff
@@ -0,0 +1,404 @@
...
+        1. It provides the predefined edges(bonds between atoms) with in a molecule.
```
Comment (mufeili) [inline python/dgl/data/qm9_edge.py:35]: "provides the predefined edges(bonds between atoms) with in a molecule." -> "includes the bonds in a molecule in the edges of the corresponding graph"

File: python/dgl/data/qm9_edge.py [hunk 8]
```diff
@@ -0,0 +1,404 @@
...
+
+
+class QM9EdgeDataset(DGLDataset):
+    r"""QM9Edge dataset for graph property prediction (regression)
+    This dataset consists of 130,831 molecules with 19 regression targets.
+    Node means atom and edge means bond.
+    
+    This dataset differs from QM9Dataset in the following points:
+        1. It provides the predefined edges(bonds between atoms) with in a molecule. 
+        2. It provides each atom with a 11d attribute and each bond with a 4d attribute.
```
Comment (mufeili) [inline python/dgl/data/qm9_edge.py:36]: Can we simply say the two datasets associate the nodes and edges with different features?

File: python/dgl/data/qm9_edge.py [hunk 9]
```diff
@@ -0,0 +1,404 @@
...
+        3. The number of regression targets is expanded to 19.
```
Comment (mufeili) [inline python/dgl/data/qm9_edge.py:37]: What are the new targets?

File: python/dgl/data/qm9_edge.py [hunk 10]
```diff
@@ -0,0 +1,404 @@
...
+
```
Comment (mufeili) [inline python/dgl/data/qm9_edge.py:38]: You should also add a note in `dgl.data.QM9Dataset` regarding the difference betweens the two datasets.

File: python/dgl/data/qm9_edge.py [hunk 11]
```diff
@@ -0,0 +1,404 @@
...
+    Reference: `"MoleculeNet: A Benchmark for Molecular Machine Learning" <https://arxiv.org/abs/1703.00564>`_
+               Atom features come from `"Neural Message Passing for Quantum Chemistry" <https://arxiv.org/abs/1704.01212>`_
```
Comment (mufeili) [inline python/dgl/data/qm9_edge.py:40]: Is `InfoGraph` the first paper that proposes this dataset variant? If so, I think you can simply cite that paper.

File: python/dgl/data/qm9_edge.py [hunk 12]
```diff
@@ -0,0 +1,404 @@
...
+    +--------+----------------------------------+-----------------------------------------------------------------------------------+---------------------------------------------+
+    | A      | :math:`A`                        | Rotational constant                                                               | :math:`\textrm{GHz}`                        |
+    +--------+----------------------------------+-----------------------------------------------------------------------------------+---------------------------------------------+
+    | B      | :math:`B`                        | Rotational constant                                                               | :math:`\textrm{GHz}`                        |
+    +--------+----------------------------------+-----------------------------------------------------------------------------------+---------------------------------------------+
+    | C      | :math:`C`                        | Rotational constant                                                               | :math:`\textrm{GHz}`                        |
+    +--------+----------------------------------+----------------------------------------
```
Comment (mufeili) [inline python/dgl/data/qm9_edge.py:98]: I guess you need more '---' to close the table.

File: python/dgl/data/qm9_edge.py [hunk 13]
```diff
@@ -0,0 +1,404 @@
...
+    Parameters
```
Comment (mufeili) [inline python/dgl/data/qm9_edge.py:99]: Add an empty line above.

File: python/dgl/data/qm9_edge.py [hunk 14]
```diff
@@ -0,0 +1,404 @@
...
+    ----------
+    load_raw: bool
```
Comment (mufeili) [inline python/dgl/data/qm9_edge.py:101]: I think this `argument` is redundant.

File: python/dgl/data/qm9_edge.py [hunk 15]
```diff
@@ -0,0 +1,404 @@
...
+        Whether to download and load raw data. Default: False, use the preprocessed data.
+    label_keys: list
+        Names of the regression property, which should be a subset of the keys in the table above.
+        If not provided, will load all the labels.
```
Comment (mufeili) [inline python/dgl/data/qm9_edge.py:105]: "will" -> "it will"

File: python/dgl/data/qm9_edge.py [hunk 16]
```diff
@@ -0,0 +1,404 @@
...
+    raw_dir : str
+        Raw file directory to download/contains the input data directory.
+        Default: ~/.dgl/
+    force_reload : bool
+        Whether to reload the dataset. Default: False
```
Comment (mufeili) [inline python/dgl/data/qm9_edge.py:110]: "False" -> "False."

File: python/dgl/data/qm9_edge.py [hunk 17]
```diff
@@ -0,0 +1,404 @@
...
+    >>> # iterate over the dataset
+    >>> for graph, labels in data:
+    ...     print(graph) # get information of each graph
+    ...     print(labels) # get labels of the corresponding graph
+    ...     # your code here...
+    >>>
+    """
+    
+    raw_url = ('https://deepchemdata.s3-us-west-1.amazonaws.com/datasets/'
+            'molnet_publish/qm9.zip')
```
Comment (mufeili) [inline python/dgl/data/qm9_edge.py:139]: Fix the indent

File: python/dgl/data/qm9_edge.py [hunk 18]
```diff
@@ -0,0 +1,404 @@
...
+        else:
+            self.targets = [self.map_dict[i] for i in label_keys]
+            self.num_labels = len(label_keys)
+            
+        self.load_raw = load_raw
+                
+        self._url = _get_dgl_url('dataset/qm9_edge.npz')
+        
+        super(QM9EdgeDataset, self).__init__(name='qm9Edge',
+                                            raw_dir=raw_dir,
```
Comment (mufeili) [inline python/dgl/data/qm9_edge.py:172]: Fix the indent

File: python/dgl/data/qm9_edge.py [hunk 19]
```diff
@@ -0,0 +1,404 @@
...
+                                            url = self._url,
```
Comment (mufeili) [inline python/dgl/data/qm9_edge.py:173]: "url = self._url" -> "url=self._url

File: python/dgl/data/qm9_edge.py [hunk 20]
```diff
@@ -0,0 +1,404 @@
...
+                                            force_reload=force_reload,
```
Comment (mufeili) [inline python/dgl/data/qm9_edge.py:174]: Fix the indent

File: python/dgl/data/qm9_edge.py [hunk 21]
```diff
@@ -0,0 +1,404 @@
...
+                                            verbose=verbose)
```
Comment (mufeili) [inline python/dgl/data/qm9_edge.py:175]: Ditto

File: python/dgl/data/qm9_edge.py [hunk 22]
```diff
@@ -0,0 +1,404 @@
...
+        if rdkit is None or not self.load_raw:
+            print('Using a pre-processed version of the dataset.'
+                  'Please install `rdkit` to alternatively process the raw data.')
+            
+            file_path = f'{self.raw_dir}/qm9_edge.npz'
+            if not os.path.exists(file_path):
+                download(self._url, path=file_path)
+    
+        else:
+            if not os.path.exists(os.path.join(self.raw_dir, "gdb9.sdf.csv")):
```
Comment (mufeili) [inline python/dgl/data/qm9_edge.py:188]: It's better to stick to the same style/usage for both branches, e.g. use one of `f'{self.raw_dir}/xxx'` and `os.path.join(self.raw_dir, "xxx")`.

File: python/dgl/data/qm9_edge.py [hunk 23]
```diff
@@ -0,0 +1,404 @@
...
+                file_path = download(self.raw_url, self.raw_dir)
+                extract_archive(file_path, self.raw_dir, overwrite=True)
+                os.unlink(file_path)
+
+            if not os.path.exists(os.path.join(self.raw_dir, "uncharacterized.txt")):
+                file_path = download(self.raw_url2, self.raw_dir)
+                os.replace(os.path.join(self.raw_dir, '3195404'),
+                        os.path.join(self.raw_dir, 'uncharacterized.txt'))
```
Comment (mufeili) [inline python/dgl/data/qm9_edge.py:196]: Fix the indent

File: python/dgl/data/qm9_edge.py [hunk 24]
```diff
@@ -0,0 +1,404 @@
...
+            self.src = data_dict['src']
+            self.dst = data_dict['dst']
+            
+            self.N_cumsum = np.concatenate([[0], np.cumsum(self.N_node)])
+            self.NE_cumsum = np.concatenate([[0], np.cumsum(self.N_edge)])
+        else:
+            with open(os.path.join(self.raw_dir, "gdb9.sdf.csv"), 'r') as f:
+                target = f.read().split('\n')[1:-1]
+                target = [[float(x) for x in line.split(',')[1:20]]
+                        for line in target]
```
Comment (mufeili) [inline python/dgl/data/qm9_edge.py:221]: Fix the indent

File: python/dgl/data/qm9_edge.py (hunk 9 above)
Comment (hengruizhang98) [reply python/dgl/data/qm9_edge.py:37]: They are listed in the tabular below

**Comment by mufeili** [comment]
In DGL, the variable names should not contain capital letters.

File: python/dgl/data/qm9_edge.py [hunk 25]
```diff
@@ -0,0 +1,404 @@
...
+                edge_attr = np.eye(len(self.bonds))[edge_type]
+
+                perm = (edge_index[0] * N + edge_index[1]).argsort()
+                edge_index = edge_index[:, perm]
+                edge_attr = edge_attr[perm]
+                
+                row, col = edge_index
+                
+                hs = (z == 1).to(F.data_type_dict['float32'])
+                out = F.zeros([int(col.max())+1], dtype=hs[row].dtype, ctx='cpu')
```
Comment (mufeili) [inline python/dgl/data/qm9_edge.py:289]: "'cpu'" -> "F.cpu()"

File: python/dgl/data/qm9_edge.py [hunk 26]
```diff
@@ -0,0 +1,404 @@
...
+                num_hs = np.array(out.scatter_add(-1, col, hs[row]))
```
Comment (mufeili) [inline python/dgl/data/qm9_edge.py:290]: I guess `out.scatter_add` will not work for all frameworks. @yzh119 Do we currently support `F.scatter_add`?

File: python/dgl/data/qm9_edge.py [hunk 27]
```diff
@@ -0,0 +1,404 @@
...
+
+                row = np.array(row)
+                col = np.array(col)
+
+                x1 = np.eye(len(self.types))[type_idx]
+                x2 = np.array([atomic_number, aromatic, sp, sp2, sp3, num_hs]).transpose()
+                x = np.concatenate((x1,x2), axis = 1)
+                
+                N_node.append(N)
+                N_edge.append(len(mol.GetBonds()) * 2)
```
Comment (mufeili) [inline python/dgl/data/qm9_edge.py:300]: "len(mol.GetBonds())" -> "mol.GetNumBonds()"

File: python/dgl/data/qm9_edge.py [hunk 28]
```diff
@@ -0,0 +1,404 @@
...
+
+            self.N_cumsum = np.concatenate([[0], np.cumsum(self.N_node)])
+            self.NE_cumsum = np.concatenate([[0], np.cumsum(self.N_edge)])
+        
+    def has_cache(self):
+        npz_path = os.path.join(self.raw_dir, "qm9_edge.npz")
+        return os.path.exists(npz_path)
+    
+    def save(self):
+        np.savez_compressed(os.path.join(self.raw_dir, "qm9_edge.npz"), N_node=self.N_node,
```
Comment (mufeili) [inline python/dgl/data/qm9_edge.py:333]: Move `N_node=self.N_node` to the next line

File: python/dgl/data/qm9_edge.py [hunk 29]
```diff
@@ -0,0 +1,404 @@
...
+                                                                        N_edge=self.N_edge,
+                                                                        Node_attr=self.Node_attr,
+                                                                        Node_pos = self.Node_pos,
+                                                                        Edge_attr = self.Edge_attr,
+                                                                        src = self.src,
+                                                                        dst = self.dst,  
+                                                                        Target = self.Target)
```
Comment (mufeili) [inline python/dgl/data/qm9_edge.py:340]: Let's be consistent. Either `x=y` or `x = y`.

File: python/dgl/data/qm9_edge.py [hunk 30]
```diff
@@ -0,0 +1,404 @@
...
+        
+        pos = self.Node_pos[self.N_cumsum[idx]:self.N_cumsum[idx + 1]]
+        src = self.src[self.NE_cumsum[idx]:self.NE_cumsum[idx+1]]
+        dst = self.dst[self.NE_cumsum[idx]:self.NE_cumsum[idx+1]]
+
+        g = dgl_graph((src, dst))
+          
+        g.ndata['pos'] = F.tensor(pos, dtype=F.data_type_dict['float32'])
+        g.ndata['attr'] = F.tensor(self.Node_attr[self.N_cumsum[idx]:self.N_cumsum[idx + 1]], dtype=F.data_type_dict['float32'])
+        g.edata['edge_attr'] = F.tensor(self.Edge_attr[self.NE_cumsum[idx]:self.NE_cumsum[idx + 1]], dtype=F.data_type_dict['float32'])
```
Comment (mufeili) [inline python/dgl/data/qm9_edge.py:388]: Either `idx+1` or `idx + 1`

File: python/dgl/data/qm9_edge.py [hunk 31]
```diff
@@ -0,0 +1,404 @@
...
+        
+        
+        label = F.tensor(self.Target[idx][self.targets], dtype=F.data_type_dict['float32'])
+        
+        return g, label
+
+    def __len__(self):
+        r"""Number of graphs in the dataset.
+        Return
```
Comment (mufeili) [inline python/dgl/data/qm9_edge.py:397]: Leave an empty line above

Comment (mufeili) [inline python/dgl/data/qm9_edge.py:397]: "Return" -> "Returns"

File: python/dgl/data/qm9_edge.py [hunk 32]
```diff
@@ -0,0 +1,403 @@
...
+conversion = F.tensor([
+    1., 1., HAR2EV, HAR2EV, HAR2EV, 1., HAR2EV, HAR2EV, HAR2EV, HAR2EV, HAR2EV,
+    1., KCALMOL2EV, KCALMOL2EV, KCALMOL2EV, KCALMOL2EV, 1., 1., 1.
+])
+
+
+class QM9EdgeDataset(DGLDataset):
+    r"""QM9Edge dataset for graph property prediction (regression)
+    This dataset consists of 130,831 molecules with 19 regression targets.
+    Nodes correspond to atoms and edges correspond to bond.
```
Comment (mufeili) [inline python/dgl/data/qm9_edge.py:35]: "bond" -> "bonds"

File: python/dgl/data/qm9_edge.py [hunk 33]
```diff
@@ -0,0 +1,403 @@
...
+    
+    This dataset differs from :class:~dgl.data.QM9Dataset in the following points:
```
Comment (mufeili) [inline python/dgl/data/qm9_edge.py:37]: Need to add "`" before "~" and "`" after "QM9Dataset"

File: python/dgl/data/qm9_edge.py (hunk 7 above)
Comment (mufeili) [reply python/dgl/data/qm9_edge.py]: Same for "~dgl.data.QM9Dataset" as above

File: python/dgl/data/qm9_edge.py (hunk 8 above)
Comment (mufeili) [reply python/dgl/data/qm9_edge.py]: Then how about "It provides edge features and a different set of node features?"

File: python/dgl/data/qm9_edge.py (hunk 11 above)
Comment (hengruizhang98) [reply python/dgl/data/qm9_edge.py:43]: The 'infograph' paper also mentioned the second reference, so we'd better not just cite 'infograph'.

File: python/dgl/data/qm9_edge.py [hunk 34]
```diff
@@ -0,0 +1,402 @@
...
+    This dataset consists of 130,831 molecules with 19 regression targets.
+    Nodes correspond to atoms and edges correspond to bonds.
+    
+    This dataset differs from :class:`~dgl.data.QM9Dataset` in the following points:
+        1. It includes the bonds in a molecule in the edges of the corresponding graph while the edges in :class:`~dgl.data.QM9Dataset` are purely distance-based.
+        2. It provides edge features, and node features in addition to the atom's locations and atomic number.
+        3. The number of regression targets is expanded from 13 to 19 (in :class:`~dgl.data.QM9Dataset`, it's 13).
+
+    Reference: `"MoleculeNet: A Benchmark for Molecular Machine Learning" <https://arxiv.org/abs/1703.00564>`_
+               Atom features come from `"Neural Message Passing for Quantum Chemistry" <https://arxiv.org/abs/1704.01212>`_
```
Comment (mufeili) [inline python/dgl/data/qm9_edge.py:43]: List references with bullet points as in "Statistics", "Node attributes" and "Edge attributes"

File: python/dgl/data/qm9_edge.py [hunk 35]
```diff
@@ -0,0 +1,402 @@
...
+
+
+class QM9EdgeDataset(DGLDataset):
+    r"""QM9Edge dataset for graph property prediction (regression)
+    This dataset consists of 130,831 molecules with 19 regression targets.
+    Nodes correspond to atoms and edges correspond to bonds.
+    
+    This dataset differs from :class:`~dgl.data.QM9Dataset` in the following points:
+        1. It includes the bonds in a molecule in the edges of the corresponding graph while the edges in :class:`~dgl.data.QM9Dataset` are purely distance-based.
+        2. It provides edge features, and node features in addition to the atom's locations and atomic number.
```
Comment (mufeili) [inline python/dgl/data/qm9_edge.py:39]: "atom's locations and atomic number" -> "atoms' locations and atomic numbers"

File: python/dgl/data/qm9_edge.py [hunk 36]
```diff
@@ -0,0 +1,402 @@
...
+    | A      | :math:`A`                        | Rotational constant                                                               | :math:`\textrm{GHz}`                        |
+    +--------+----------------------------------+-----------------------------------------------------------------------------------+---------------------------------------------+
+    | B      | :math:`B`                        | Rotational constant                                                               | :math:`\textrm{GHz}`                        |
+    +--------+----------------------------------+-----------------------------------------------------------------------------------+---------------------------------------------+
+    | C      | :math:`C`                        | Rotational constant                                                               | :math:`\textrm{GHz}`                        |
+    +--------+----------------------------------+----------------------------------------------------------------------------------------------------------------------------------
```
Comment (mufeili) [inline python/dgl/data/qm9_edge.py:101]: The last one should be "+".

File: python/dgl/data/qm9_edge.py [hunk 37]
```diff
@@ -0,0 +1,402 @@
...
+    
+        else:
+            if not os.path.exists(f'{self.raw_dir}/gdb9.sdf.csv'):
+                file_path = download(self.raw_url, self.raw_dir)
+                extract_archive(file_path, self.raw_dir, overwrite=True)
+                os.unlink(file_path)
+
+            if not os.path.exists(f'{self.raw_url2}/uncharacterized.txt'):
+                file_path = download(self.raw_url2, self.raw_dir)
+                os.replace(os.path.join(self.raw_dir, '3195404'), os.path.join(self.raw_dir, 'uncharacterized.txt'))
```
Comment (mufeili) [inline python/dgl/data/qm9_edge.py:194]: Use `f'{self.raw_dir}/...'` for consistency.

File: python/dgl/data/qm9_edge.py [hunk 38]
```diff
@@ -0,0 +1,402 @@
...
+                edge_attr = np.eye(len(self.bonds))[edge_type]
+
+                perm = (edge_index[0] * N + edge_index[1]).argsort()
+                edge_index = edge_index[:, perm]
+                edge_attr = edge_attr[perm]
+                
+                row, col = edge_index
+                
+                hs = (z == 1).to(F.data_type_dict['float32'])
+                out = F.zeros([int(col.max())+1], dtype=hs[row].dtype, ctx= F.cpu())
```
Comment (mufeili) [inline python/dgl/data/qm9_edge.py:285]: `ctx= F.cpu()` -> `ctx=F.cpu()`

File: python/dgl/data/qm9_edge.py (hunk 26 above)
Comment (yzh119) [reply python/dgl/data/qm9_edge.py:288]: Yes we do, @hengruizhang98 could you please explain what is `out` and how does `hs`/`num_hs` work?

**Comment by mufeili** [comment]
I talked to Minjie. He suggested moving the code for preprocessing to a gist file and leave a link to the gist file.

**Comment by hengruizhang98** [comment]
> I talked to Minjie. He suggested moving the code for preprocessing to a gist file and leave a link to the gist file.

Now the preprocessing part has been moved to : https://gist.github.com/hengruizhang98/a2da30213b2356fff18b25385c9d3cd2

File: python/dgl/data/qm9_edge.py (hunk 26 above)
Comment (hengruizhang98) [reply python/dgl/data/qm9_edge.py]: num_hs[i] indicates the number of hydrogen atoms connected to node[i].
I've rewritten it with numpy:

hs = (np.array(atomic_number) == 1).astype(np.int64)

num_hs = np.zeros(n_atom)
for i in range(col.shape[0]):
    num_hs[col[i]] += hs[row][i]

Comment (yzh119) [reply python/dgl/data/qm9_edge.py]: So the feature size is one, and we have an API `scatter_add` for that:
https://github.com/dmlc/dgl/blob/master/python/dgl/backend/backend.py#L1555-L1575

Comment (hengruizhang98) [reply python/dgl/data/qm9_edge.py]: Thanks! Now it has been changed into: 

x = F.tensor(hs[row], dtype=F.data_type_dict['float32'])
idx = F.tensor(col, dtype=F.data_type_dict['int64'])
num_hs = F.scatter_add(x, idx, n_atom)

File: python/dgl/data/qm9.py [hunk 39]
```diff
@@ -15,6 +15,11 @@ class QM9Dataset(DGLDataset):
     This dataset consists of 13,0831 molecules with 12 regression targets.
     Node means atom and edge means bond.
```
Comment (mufeili) [inline python/dgl/data/qm9.py:16]: Nodes correspond to atoms and edges correspond to bonds.

File: python/dgl/data/qm9.py [hunk 40]
```diff
@@ -15,6 +15,11 @@ class QM9Dataset(DGLDataset):
     This dataset consists of 13,0831 molecules with 12 regression targets.
```
Comment (mufeili) [inline python/dgl/data/qm9.py:15]: "13,0831" -> "130,831"

File: python/dgl/data/qm9.py [hunk 41]
```diff
@@ -15,6 +15,11 @@ class QM9Dataset(DGLDataset):
...
     Node means atom and edge means bond.
 
+    This dataset differers from :class:`~dgl.data.QM9EdgeDataset` in the following aspects:
```
Comment (mufeili) [inline python/dgl/data/qm9.py:18]: "differers" -> "differs"

File: python/dgl/data/qm9.py [hunk 42]
```diff
@@ -15,6 +15,11 @@ class QM9Dataset(DGLDataset):
...
+        1. Edges in this dataset are purely distance-based
```
Comment (mufeili) [inline python/dgl/data/qm9.py:19]: "distance-based" -> "distance-based."

File: python/dgl/data/qm9.py [hunk 43]
```diff
@@ -15,6 +15,11 @@ class QM9Dataset(DGLDataset):
...
+        2. It only provide the atoms' coordinates and atomic numvers as node attributes.
```
Comment (mufeili) [inline python/dgl/data/qm9.py:20]: It only provides atoms' coordinates and atomic numbers as node features.

File: python/dgl/data/qm9.py [hunk 44]
```diff
@@ -15,6 +15,11 @@ class QM9Dataset(DGLDataset):
     This dataset consists of 13,0831 molecules with 12 regression targets.
     Node means atom and edge means bond.
 
+    This dataset differers from :class:`~dgl.data.QM9EdgeDataset` in the following aspects:
+        1. Edges in this dataset are purely distance-based
+        2. It only provide the atoms' coordinates and atomic numvers as node attributes. 
+        3. It only provide 12 regression targets.
```
Comment (mufeili) [inline python/dgl/data/qm9.py:21]: It only provides 12 regression targets.

File: python/dgl/data/qm9_edge.py (hunk 26 above)
Comment (yzh119) [reply python/dgl/data/qm9_edge.py]: I'm okay with about change.

**Review by mufeili — APPROVED** [review]
LGTM.

**Pull request merged**